mithril-stm/src/stm.rs

```diff
@@ -705,35 +779,35 @@ impl<D: Clone + Digest + FixedOutput + Send + Sync> StmAggrSig<D> {
         msg: &[u8],
         avk: &StmAggrVerificationKey<D>,
         parameters: &StmParameters,
-    ) -> Result<(), StmAggregateSignatureError<D>> {
-        let mut nr_indices = 0;
-        let mut unique_indices = HashSet::new();
-        let mut leaves = Vec::new();
-
-        for (sig, reg_party) in &self.signatures {
-            let msgp = avk.mt_commitment.concat_with_msg(msg);
-            sig.check_indices(parameters, &reg_party.1, &msgp, avk)?;
-
-            for &index in &sig.indexes {
-                unique_indices.insert(index);
-                nr_indices += 1;
-            }
+    ) -> Result<(Vec<Signature>, Vec<VerificationKey>), StmAggregateSignatureError<D>> {
+        let msgp = avk.mt_commitment.concat_with_msg(msg);
 
-            leaves.push(*reg_party);
-        }
+        let leaves = self
+            .signatures
+            .iter()
+            .map(|r| r.1)
+            .collect::<Vec<RegParty>>();
 
-        if nr_indices != unique_indices.len() {
-            return Err(StmAggregateSignatureError::IndexNotUnique);
-        }
+        let signatures = self
+            .signatures
+            .iter()
+            .map(|r| r.0.clone())
+            .collect::<Vec<StmSig>>();
 
-        if (nr_indices as u64) < parameters.k {
-            return Err(StmAggregateSignatureError::NoQuorum);
-        }
+        FullNodeVerifier::pre_verify(
+            &avk.total_stake,
+            signatures.as_slice(),
+            parameters,
+            msgp.as_slice(),
+            leaves.clone(),
```

Comment:
```suggestion
            &avk.total_stake,
            &signatures,
            parameters,
            &msgp,
            leaves.clone(),
```
we probably don't want to clone leaves. See comment below.